Bot.Core/TownMaintenance.cs

```diff
@@ -0,0 +1,106 @@
+﻿using Bot.Api;
+using Bot.Api.Database;
+using Bot.Core.Callbacks;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Bot.Core
+{
+    internal class TownMaintenance : ITownMaintenance
+    {
+        const int NUM_TOWNS_PER_CALLBACK = 5;
+        const int MINUTES_PER_CALLBACK = 5;
+
+        private readonly ICallbackScheduler<Queue<TownKey>> m_callbackScheduler;
+        private readonly ITownDatabase m_townDatabase;
+        private readonly IBotClient m_botClient;
+        private readonly IDateTime m_dateTime;
+        private readonly IShutdownPreventionService m_shutdownPrevention;
+
+        private readonly TaskCompletionSource m_shutdownTcs = new();
+        private bool m_shutdownRequested = false;
+        private bool m_queueProcessing = false;
+
+        List<Func<TownKey, Task>> m_startupTasks = new();
+
+        public TownMaintenance(IServiceProvider sp)
+        {
+            sp.Inject(out m_dateTime);
+            sp.Inject(out m_townDatabase);
+            sp.Inject(out m_botClient);
+            sp.Inject(out m_shutdownPrevention);
+
+            m_shutdownPrevention.ShutdownRequested += (s, e) =>
+            {
+                m_shutdownRequested = true;
+                if (!m_queueProcessing)
+                {
+                    m_shutdownTcs.TrySetResult();
+                }
+            };
+            m_shutdownPrevention.RegisterShutdownPreventer(m_shutdownTcs.Task);
+
+            var callbackFactory = sp.GetService<ICallbackSchedulerFactory>();
+            m_callbackScheduler = callbackFactory.CreateScheduler<Queue<TownKey>>(ProcessQueue, TimeSpan.FromMinutes(1));
+
+            m_botClient.Connected += async (o, args) => await RunMaintenanceAsync();
+        }
+
+        private async Task ProcessQueue(Queue<TownKey> towns)
+        {
+            m_queueProcessing = true;
```

Comment:
I think this should start with 
```
if (m_queueProcessing)
   return;
```

That way if the queue was already processing, it would not start another process.